internals/daemon/daemon.go

```diff
@@ -47,13 +45,40 @@ import (
 	"github.com/canonical/pebble/internals/systemd"
 )
 
+// notifySupervisor interface defines the required API for
+// sending notifications to a supervisor daemon.
+type notifySupervisor interface {
+	Available() bool
+	Send(state string) error
+}
+
+// shutdown defines the required API for reboot, poweroff
+// and halt control of the underlying system
+type shutdown interface {
+	Reboot(delay time.Duration, msg string) error
+}
+
 var (
 	ErrRestartSocket = fmt.Errorf("daemon stop requested to wait for socket activation")
 
-	systemdSdNotify = systemd.SdNotify
-	sysGetuid       = sys.Getuid
+	sysGetuid = sys.Getuid
+
+	// The following functions should be modified depending on the
+	// underlying system in use. For example, some systems may
+	// not have systemd, and will use this system manager directly.
+
+	// Notify provides a way to send a notification to a supervisor
+	// daemon, if one exists. The default implementation assumes
+	// systemd and should be replaced if not available.
+	Notify notifySupervisor = systemd.Notifier
+
+	// Shutdown provides the system specific functionality to
+	// halt, poweroff or reboot.
+	Shutdown shutdown = systemd.Shutdown
```

Comment:
It seems really unfortunate that these need to be globals when the `Daemon` is created via the more typical `daemon.New` pattern (that is, not a global). Can these somehow be part of the `Options` struct and not be globals of the daemon package? But I guess the tricky part is then having the alternative CLI customize it.

One approach would be to add Notify and Shutdown to `daemon.Options`, but then have a `ConfigureDaemon` that's part of the CLI's "personality". The signature would be `ConfigureDaemon func(opts *daemon.Options)`, and the alternative CLI would do something like:

```go
cmd.Personality.ConfigureDaemon = func(opts *daemon.Options) {
    opts.Notify = mySpecialNotify
    opts.Shutdown = mySpecialShutdown
}
```

I dunno, maybe it's not any better, but at least it would confine the messy globals to one place, the "personality" struct. Or we could put the globals in the `internals/cli` package -- putting them in the CLI still seems better to me than in the daemon package.

Reply:
I like the idea of confinement but is this really a part of the Personality, I can see this behaviour growing with other options, so a new struct to embody them?

Reply:
I don't think this is part of Personality, because (i) it does not concern other components apart from the daemon, and (ii) Termus *will* have its own initialization of the daemon during bootstrap.

I'm on board @benhoyt's idea of having those be inside `daemon.Options` 😃